pkg/cluster/streams.go

```diff
@@ -0,0 +1,261 @@
+
+func (c *Cluster) createStreams() error {
+	c.setProcessName("creating streams")
+
+	fes := c.generateFabricEventStream()
+	_, err := c.KubeClient.FabricEventStreams(c.Namespace).Create(context.TODO(), fes, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("could not create event stream custom resource: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Cluster) updateStreams(newEventStreams *zalandov1alpha1.FabricEventStream) error {
+	c.setProcessName("updating event streams")
+
+	_, err := c.KubeClient.FabricEventStreams(newEventStreams.Namespace).Update(context.TODO(), newEventStreams, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("could not update event stream custom resource: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Cluster) deleteStreams() error {
+	c.setProcessName("deleting event streams")
+
+	// check if stream CRD is installed before trying a delete
+	_, err := c.KubeClient.CustomResourceDefinitions().Get(context.TODO(), constants.EventStreamSourceCRDName, metav1.GetOptions{})
+	if k8sutil.ResourceNotFound(err) {
+		return nil
+	}
+
+	err = c.KubeClient.FabricEventStreams(c.Namespace).Delete(context.TODO(), c.Name, metav1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("could not delete event stream custom resource: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Cluster) syncPostgresConfig() error {
+
+	desiredPostgresConfig := make(map[string]interface{})
+	slots := make(map[string]map[string]string)
+
+	// if streams are defined wal_level must be switched to logical and slots have to be defined
+	desiredPostgresConfig["postgresql"] = map[string]interface{}{patroniPGParametersParameterName: map[string]string{"wal_level": "logical"}}
+
+	for _, stream := range c.Spec.Streams {
+		slotName := c.getLogicalReplicationSlot(stream.Database)
+
+		if slotName == "" {
+			slot := map[string]string{
+				"database": stream.Database,
+				"plugin":   "wal2json",
+				"type":     "logical",
+			}
+			slots[constants.EventStreamSourceSlotPrefix+"_"+stream.Database] = slot
+		}
+	}
+
+	if len(slots) > 0 {
+		c.logger.Debugf("setting wal level to 'logical' in Postgres configuration")
+		for slotName, slot := range slots {
+			c.logger.Debugf("creating logical replication slot %q in database %q", slotName, slot["database"])
+		}
+		desiredPostgresConfig["slots"] = slots
+	} else {
+		return nil
+	}
+
+	pods, err := c.listPods()
+	if err != nil || len(pods) == 0 {
+		return err
+	}
+	for i, pod := range pods {
+		podName := util.NameFromMeta(pods[i].ObjectMeta)
+		effectivePostgresConfig, err := c.patroni.GetConfig(&pod)
+		if err != nil {
+			c.logger.Warningf("could not get Postgres config from pod %s: %v", podName, err)
+			continue
+		}
+
+		_, err = c.checkAndSetGlobalPostgreSQLConfiguration(&pod, effectivePostgresConfig, desiredPostgresConfig)
+		if err != nil {
+			c.logger.Warningf("could not set PostgreSQL configuration options for pod %s: %v", podName, err)
```

Comment:
Just to double check: here it is assumed the operator will retry setting the `effectivePostgresConfig` on the next Sync, right ?

Reply:
Yes